src/optimizer/agg_key_dependency_optimizer.cpp

```diff
@@ -0,0 +1,76 @@
+#include "optimizer/agg_key_dependency_optimizer.h"
+
+#include "binder/expression/property_expression.h"
+#include "planner/logical_plan/logical_operator/logical_aggregate.h"
+#include "planner/logical_plan/logical_operator/logical_distinct.h"
+
+using namespace kuzu::planner;
+
+namespace kuzu {
+namespace optimizer {
+
+void AggKeyDependencyOptimizer::rewrite(planner::LogicalPlan* plan) {
+    visitOperator(plan->getLastOperator().get());
+}
+
+void AggKeyDependencyOptimizer::visitOperator(planner::LogicalOperator* op) {
+    // bottom up traversal
+    for (auto i = 0u; i < op->getNumChildren(); ++i) {
+        visitOperator(op->getChild(i).get());
+    }
+    visitOperatorSwitch(op);
+}
+
+void AggKeyDependencyOptimizer::visitAggregate(planner::LogicalOperator* op) {
+    auto agg = (LogicalAggregate*)op;
+    auto [keyExpressions, payloadExpressions] = resolveKeysAndPayloads(agg->getKeyExpressions());
+    agg->setKeyExpressions(keyExpressions);
+    agg->setPayloadExpressions(payloadExpressions);
+}
+
+void AggKeyDependencyOptimizer::visitDistinct(planner::LogicalOperator* op) {
+    auto distinct = (LogicalDistinct*)op;
+    auto [keyExpressions, payloadExpressions] =
+        resolveKeysAndPayloads(distinct->getKeyExpressions());
+    distinct->setKeyExpressions(keyExpressions);
+    distinct->setPayloadExpressions(payloadExpressions);
+}
+
+std::pair<binder::expression_vector, binder::expression_vector>
+AggKeyDependencyOptimizer::resolveKeysAndPayloads(const binder::expression_vector& keys) {
+    // Collect primary keys from group keys
+    std::vector<binder::PropertyExpression*> primaryKeys;
+    for (auto& expression : keys) {
+        if (expression->expressionType == common::PROPERTY) {
+            auto propertyExpression = (binder::PropertyExpression*)expression.get();
+            if (propertyExpression->isPrimaryKey() || propertyExpression->isInternalID()) {
+                primaryKeys.push_back(propertyExpression);
+            }
+        }
+    }
+    // Collect variable names whose primary key is part of group keys
+    std::unordered_set<std::string> primaryVarNames;
```

Comment:
I think if you gave an example, as we discussed this could be a bit more readable. preimaryKeys vs primaryVarNames and in general varNames vs expressions was not directly clear (though the code is clear enough so I could predict that's what was happening).